src/context/PubNubContext.tsx

```diff
@@ -1,19 +1,38 @@
 import React from 'react';
+import PubNub from 'pubnub';
+
+import invariant from '../invariant';
 
 export interface PubNubContextValue {
-  client?: object;
-  renderPromises?: Record<any, any>;
+  client: PubNub;
 }
 
-let pubNubContext: React.Context<PubNubContextValue>;
+export const PubNubContext = React.createContext<PubNubContextValue | null>(
+  null
+);
 
-export function getPubNubContext() {
-  if (!pubNubContext) {
-    pubNubContext = React.createContext<PubNubContextValue>({});
-  }
-  return pubNubContext;
+export interface PubNubProviderProps<T> {
+  client: T;
+  children: React.ReactNode | React.ReactNode[] | null;
 }
 
-export function resetPubNubContext() {
-  pubNubContext = React.createContext<PubNubContextValue>({});
-}
+export const PubNubProvider: React.FC<PubNubProviderProps<PubNub>> = ({
```

Comment:
Is there a reason we're including both Provider and Context in the same file? I think the intent of splitting them was for separation of concerns for possible scalability changes.

Reply:
So: the Provider and Context are very tightly coupled, so it doesn't really make sense to separate them. Also keep in mind that our Provider is just a very thin wrapper around the 
default React provider. We can separate them into different files, but just based on my experience it is not really worth it at this point.

Reply:
I don't have a strong opinion, but when I look around at other libraries, they do split up provider/context/consumer into separate modules.